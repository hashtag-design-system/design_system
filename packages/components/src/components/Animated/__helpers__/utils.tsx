import { screen } from "@testing-library/react";

export const checkChildStyle = (element: string | HTMLElement, styleValue: { name: string; value: React.ReactText }) => {
  const { name, value } = styleValue;
  if (typeof element === "string") {
    expect(screen.getByTestId(element).style["_values"][`--${name}`]).toBe(String(value));
  } else {
    expect(element.style["_values"][`--${name}`]).toBe(String(value));
  }
};

export const checkStyleCustomProperties = (element: HTMLElement, arr: string[]) => {
  arr
    .map(item => `--${item}`)
    .forEach((style, i) => {
      expect(element.style[i.toString()]).toBe(style);
    });
};
