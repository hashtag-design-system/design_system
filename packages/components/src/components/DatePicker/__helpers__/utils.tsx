import dayjs from "dayjs";
import React from "react";
import Select from "../../Select";
import DatePicker, { DatePickerFProps, DatePickerOnChangeInfo } from "../index";

type FProps = DatePickerFProps;

export const TEST_DEFAULT_DATE = dayjs("2021-01-26");

export const DefaultDatePicker: React.FC<FProps> = ({ ...props }) => {
  return <DatePicker defaultOpen defaultDates={[TEST_DEFAULT_DATE]} defaultCalendarDate={TEST_DEFAULT_DATE} {...props} />;
};

export const TestDatePickerButton: React.FC<Pick<DatePickerOnChangeInfo, "selectedDate">> = ({ selectedDate }) => (
  <Select.Button>{selectedDate.length >= 1 ? selectedDate[0].format("DD/MM/YYYY") : ""}</Select.Button>
);
