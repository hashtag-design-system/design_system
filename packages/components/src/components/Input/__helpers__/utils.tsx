import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { InputContextProvider } from "../../../utils/contexts/InputContext";
import { generateInputId } from "../../../utils/hooks/useInputId";
import { InputFProps } from "../index";

export const defaultProps: InputFProps = {
  id: generateInputId(),
  placeholder: "Placeholder",
  floatingplaceholder: true,
  state: "default",
};

type InputCustomRenderOptions = {
  providerProps: InputFProps;
  renderOptions?: RenderOptions;
};

export const inputCustomRender = (ui: React.ReactElement, options?: InputCustomRenderOptions) => {
  return render(
    <InputContextProvider value={{ ...defaultProps, ...options?.providerProps }}>{ui}</InputContextProvider>,
    options?.renderOptions
  );
};
