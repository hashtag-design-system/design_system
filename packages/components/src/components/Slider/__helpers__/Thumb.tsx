import React, { useState } from "react";
import { useSliderContext } from "../../../utils/contexts/SliderContext";

type Props = {
  size: number;
};

export const Thumb: React.FunctionComponent<Props> = ({ size }) => {
  const [prevKey, setPrevKey] = useState<string>("0");

  const { value, setValue, calcValue, calcPercentage, max, step } = useSliderContext();

  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    const { key } = e;

    if (key === "ArrowRight" || key === "ArrowUp") {
      setValue(value => value + step);
    } else if (key === "ArrowLeft" || key === "ArrowDown") {
      setValue(value => value - step);
    }

    if (!isNaN(+key)) {
      const numKey = parseInt(key);
      const number = calcValue(numKey * 10);
      if (numKey === 0 && prevKey === "1") {
        setValue(max);
      } else {
        setValue(number);
      }
    }

    setPrevKey(key);
  };

  return (
    <div
      className="slider__thumb flex-row-center-center"
      style={{
        left: `${calcPercentage()}%`,
        width: `${size}em`,
        height: `${size}em`,
      }}
      tabIndex={0}
      onKeyDown={e => handleKeyDown(e)}
    >
      <span className="slider__thumb__value body-16">{value}</span>
    </div>
  );
};
