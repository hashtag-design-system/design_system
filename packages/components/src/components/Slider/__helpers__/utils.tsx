import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { SliderContextProvider } from "../../../utils/contexts/SliderContext";
import { SliderFProps } from "../index";

export const defaultProps: SliderFProps = {
  min: 0,
  max: 100,
  step: 1,
  marks: [
    { value: 0, label: "0" },
    { value: 100, label: "100" },
  ],
  lockOnMarks: false,
  zeroPercentageOnEdgeMarks: false,
};

type SliderCustomRenderOptions = {
  providerProps: SliderFProps;
  renderOptions?: RenderOptions;
};

export const sliderCustomRender = (ui: React.ReactElement, options?: SliderCustomRenderOptions) => {
  return render(
    <SliderContextProvider value={{ ...defaultProps, ...options?.providerProps, calcPercentage: () => 1, calcValue: () => 1 }}>
      {ui}
    </SliderContextProvider>,
    options?.renderOptions
  );
};
