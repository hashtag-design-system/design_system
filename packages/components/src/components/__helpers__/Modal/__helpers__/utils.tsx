import { screen, waitFor } from "@testing-library/react";
import { CONFIG } from "../../../../config";

export const overlayCheckStyle = async (toContain: string): Promise<void> => {
  const modal = screen.getByTestId("modal");

  // Wait due to animation from `opacity: 0` to `opacity: 1`
  await waitFor(() => {
    expect(modal).toBeVisible();
  });

  expect(modal).toHaveAttribute("style");
  if (toContain.includes("rgba")) {
    expect(modal.style.backgroundColor).toContain(toContain);
  } else {
    expect(modal.style["backdropFilter"]).toContain(toContain);
  }
};

export const createPortalElement = () => {
  const newEl = document.createElement("div");
  newEl.setAttribute("id", CONFIG.PORTAL_ID_SELECTOR);
  newEl.setAttribute("data-testid", CONFIG.PORTAL_ID_SELECTOR);
  document.body.prepend(newEl);
  return newEl;
};