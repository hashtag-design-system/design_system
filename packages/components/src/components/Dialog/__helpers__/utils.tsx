import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { DialogContextProvider, DialogContextType } from "../../../utils/contexts";

export const defaultProps: Omit<DialogContextType, "handleDismiss"> = {
  hasBtnGroup: false,
  confirm: false,
  loading: false,
  allowDismissOnLoading: true,
};

type DialogCustomRenderOptions = {
  providerProps: Partial<DialogContextType>;
  renderOptions?: RenderOptions;
};

export const dialogCustomRender = (ui: React.ReactElement, options?: DialogCustomRenderOptions) => {
  return render(
    // @ts-expect-error
    <DialogContextProvider value={{ ...defaultProps, ...options?.providerProps }}>{ui}</DialogContextProvider>,
    options?.renderOptions
  );
};
