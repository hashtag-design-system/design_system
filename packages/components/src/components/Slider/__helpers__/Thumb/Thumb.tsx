import React from "react";
import { useSliderContext } from "../../../../utils";
import { ComponentProps } from "../../../__helpers__";
import { SliderFProps } from "../../index";

type Props = {
  size: number;
  value: number;
  onHover?: boolean;
  focusVisible?: boolean;
} & Pick<ComponentProps<"div">, "onMouseOver">;

export type FProps = Props & Required<Pick<SliderFProps, "thumb">> & Pick<SliderFProps, "className">;

export const Thumb: React.FunctionComponent<FProps> = ({
  onHover = false,
  size,
  thumb,
  value,
  focusVisible,
  className,
  onMouseOver,
}) => {
  const { formatRegExp } = thumb;
  const { max, calcPercentage } = useSliderContext();
  const newVal = className?.includes("right") && max !== undefined ? parseFloat(max.toString()) - value : value;

  return (
    <div
      className={`slider__thumb ${className ? className : ""} ${focusVisible ? "focus-visible" : ""} ${onHover ? "hover" : ""}`}
      style={{
        [className || "left"]: `${calcPercentage(value)}%`,
        width: `${size}em`,
        height: `${size}em`,
      }}
      tabIndex={0}
      data-testid="slider-thumb"
      onMouseOver={onMouseOver}
    >
      <span className="slider__thumb__value body-16" data-testid="slider-thumb-value">
        {formatRegExp ? String(newVal).replace(formatRegExp.searchValue, formatRegExp.replaceValue) : newVal}
      </span>
    </div>
  );
};
