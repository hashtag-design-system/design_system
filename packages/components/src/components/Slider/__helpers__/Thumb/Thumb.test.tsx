import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { defaultProps, sliderCustomRender } from "../utils";
import { FProps as ThumbFProps, Thumb } from "./Thumb";

const TEST_COMPONENT: React.FunctionComponent<Partial<ThumbFProps>> = ({
  thumb = { defaultValue: 10 },
  size = 24,
  value = 30,
  ...props
}) => {
  return <Thumb size={size} value={value} thumb={thumb} {...props} />;
};

describe("Slider <Thumb />", () => {
  test("default behaviour", () => {
    sliderCustomRender(<TEST_COMPONENT />);
    const thumb = screen.getByTestId("slider-thumb");
    const thumbValue = screen.getByTestId("slider-thumb-value");

    expect(thumb).toBeVisible();
    expect(thumb.children).toHaveLength(1);
    expect(thumb).toMatchSnapshot();
    expect(thumb).toHaveAttribute("class");
    expect(thumb).toHaveAttribute("tabindex", "0");
    expect(thumb.style.width).toContain("24");
    expect(thumb.style.height).toContain("24");
    expect(thumb.style.left).toBeDefined();
    expect(thumbValue).toBeVisible();
    expect(thumbValue).toHaveAttribute("class");
    expect(thumbValue).toHaveTextContent("30");
  });
  test("with onHover={true}", () => {
    sliderCustomRender(<TEST_COMPONENT onHover />);

    expect(screen.getByTestId("slider-thumb")).toHaveClass("hover");
  });
  test("with focusVisible={true}", () => {
    sliderCustomRender(<TEST_COMPONENT focusVisible />);

    expect(screen.getByTestId("slider-thumb")).toHaveClass("focus-visible");
  });
  test("with custom className", () => {
    sliderCustomRender(<TEST_COMPONENT className="right" />);
    const thumb = screen.getByTestId("slider-thumb");

    expect(thumb).toHaveClass("right");
    expect(thumb.style.right).toBeDefined();
    expect(thumb.style.left).toBe("");
    expect(screen.getByTestId("slider-thumb-value")).toHaveTextContent(String(parseFloat(defaultProps.max!.toString()) - 30));
  });
  test("with custom className", () => {
    sliderCustomRender(
      <TEST_COMPONENT value={10} thumb={{ defaultValue: 10, formatRegExp: { searchValue: new RegExp(/0/), replaceValue: "" } }} />
    );

    expect(screen.getByTestId("slider-thumb-value")).toHaveTextContent("1");
    expect(screen.getByTestId("slider-thumb")).toMatchSnapshot();
  });
  test("with onMouserOver custom event", () => {
    const onMouseOver = jest.fn((e: React.MouseEvent<HTMLDivElement, MouseEvent>) => e.currentTarget.dataset);
    sliderCustomRender(<TEST_COMPONENT onMouseOver={e => onMouseOver(e)} />);
    const thumb = screen.getByTestId("slider-thumb");

    expect(thumb.onmouseover).toBeDefined();

    userEvent.hover(thumb);

    expect(onMouseOver).toHaveBeenCalledTimes(1);
    expect(onMouseOver.mock.results[0].value["testid"]).toBe("slider-thumb");
  });
});
