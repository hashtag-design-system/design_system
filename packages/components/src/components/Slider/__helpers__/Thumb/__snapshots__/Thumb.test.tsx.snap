// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider <Thumb /> default behaviour 1`] = `
<div
  class="slider__thumb   "
  data-testid="slider-thumb"
  style="left: 1%; width: 24em; height: 24em;"
  tabindex="0"
>
  <span
    class="slider__thumb__value body-16"
    data-testid="slider-thumb-value"
  >
    30
  </span>
</div>
`;

exports[`Slider <Thumb /> with custom className 1`] = `
<div
  class="slider__thumb   "
  data-testid="slider-thumb"
  style="left: 1%; width: 24em; height: 24em;"
  tabindex="0"
>
  <span
    class="slider__thumb__value body-16"
    data-testid="slider-thumb-value"
  >
    1
  </span>
</div>
`;
